internal/server/data/migrations.go

```diff
@@ -1173,3 +1174,70 @@ func addUserPublicKeyUserIDIndex() *migrator.Migration {
 		},
 	}
 }
+
+func addGrantsSubjectID() *migrator.Migration {
+	return &migrator.Migration{
+		ID: "2023-01-12T17:00",
+		Migrate: func(tx migrator.DB) error {
+			if !migrator.HasColumn(tx, "grants", "subject") {
+				return nil
+			}
+
+			// Step 1 - Add the new columns
+			_, err := tx.Exec(`
+				ALTER TABLE grants ADD COLUMN IF NOT EXISTS subject_id bigint;
+				ALTER TABLE grants ADD COLUMN IF NOT EXISTS subject_kind smallint;
+			`)
+			if err != nil {
+				return err
+			}
+
+			// Step 2- Migrate the data
+			type grantSubject struct {
+				ID              uid.ID
+				OriginalSubject string
+			}
+
+			rows, err := tx.Query(`SELECT id, subject FROM grants`)
+			if err != nil {
+				return err
+			}
+			grants, err := scanRows(rows, func(g *grantSubject) []any {
+				return []any{&g.ID, &g.OriginalSubject}
+			})
+			if err != nil {
+				return err
+			}
+
+			for _, grant := range grants {
+				var kind int
+				var subjectID uid.ID
+				if strings.HasPrefix(grant.OriginalSubject, "i:") {
+					kind = 1
+				} else {
+					kind = 2
+				}
+				subjectID, err = uid.Parse([]byte(grant.OriginalSubject[2:]))
+				if err != nil {
+					return err
+				}
+
+				_, err = tx.Exec(`UPDATE grants SET subject_id=?, subject_kind=? WHERE id = ?`,
+					subjectID, kind, grant.ID)
+				if err != nil {
+					return err
+				}
+			}
+
+			// Step 3 - Remove the old column, add a new index, and NOT NULL
+			_, err = tx.Exec(`
+				ALTER TABLE grants DROP COLUMN IF EXISTS subject;
+				ALTER TABLE grants ALTER COLUMN subject_id SET NOT NULL;
+				ALTER TABLE grants ALTER COLUMN subject_kind SET NOT NULL;
```

Comment:
This final step in the migration will cause problems for the other pods (that have not yet been updated to the new code). We could save this step for a later commit, or we could accept the few errors while the pods restart.

Saving it for a later commit only works for SaaS. We'd have to spread this out over 2 releases for on-prem (or accept a few errors during upgrade).

Reply:
A few errors in dying pods is likely fine. It doesn't make sense to delay this switchover to mitigate a few seconds of errors